# Import Libraries 

In [1]:
#Import libraries 
import pandas as pd
import numpy as np
import os

# Import Data 

In [2]:
#Set Path 
path= r'C:\Users\hmaio\Documents\4.0 Instacart Project'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '4.0 Data', 'Original Data', 'products.csv'), index_col = False) 

In [4]:
df_ords = pd.read_csv(os.path.join(path, '4.0 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False) 

# Exercise Work + Answer To Task Question 1. 

In [17]:
df_ords.describe()

Unnamed: 0      order_id   customer_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [18]:
#renaming customer_id back to user_id 
df_ords.rename(columns = {'customer_id' : 'user_id'}, inplace = True)

In [19]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [20]:
# creating a dataframe 
df_test = pd.DataFrame()

In [21]:
# create mixed type column 
df_test['mix'] = ['a','b','1', 'True']

In [22]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [23]:
# check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [24]:
#changing mix column to string 
df_test['mix'] = df_test['mix'].astype('str')


In [25]:
#checking for missing values in df.prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [27]:
#viewing missing data is df_prods 
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [28]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [29]:
# creating a new df that omits missing values in product_name column 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [30]:
# looking for duplicates in df_prods 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [31]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [32]:
#removing duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

# 4.5 Task Answers for Questions 2-9

In [34]:
#answer2 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max of prices seems to be an error as it is abnormally high. 

In [35]:
# answer 3 checking for mixed data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len (df_test[weird]) > 0:
C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len (df_test[weird]) > 0:
C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len (df_test[weird]) > 0:
C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len (df_test[weird]) > 0:
C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if len (df_test[weird]) > 0:
C:\Users\hmaio\AppData\Local\Temp\ipykernel_3656\941762342.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.

Answer 4 
It would appear that non of the columns have a mixed data type 

In [37]:
# Answer 5 checking for missing values 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Answer 5 : There appears to be 206209 missing values in the days_since_prior_order column 

In [38]:
# Answer 6 creating df_nan_1 to view missing values 
df_nan_1= df_ords[df_ords['days_since_prior_order'].isnull() == True]


In [39]:
df_nan_1

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [40]:
# Answer 6 creating a new data frame to exclude the missing values 
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

Answer 6 : I decided to create a new dataframe as there was a large amount of rows with missing values and they could be important to another department. 

In [41]:
#Answer 7 checking for duplicates in df_ords_clean 
df_dups_2 = df_ords_clean [ df_ords_clean.duplicated()]

In [42]:
df_dups_2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Answer 7 There appear to be no duplicates in df_ords_clean 

In [43]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
5                 5   3367565        1             6                   2   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
5                        7                    19.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3214874 rows x 7 columns]

There appear to be no duplicates in this dataset. If there were I would have created a new dataframe and used the drop duplicates function  

In [45]:
#Answer 9 Exporting df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '4.0 Data','Prepared Data', 'products_checked.csv'))

In [46]:
#Answer 9 Exporting df_ords_clean
df_ords_clean.to_csv(os.path.join(path, '4.0 Data','Prepared Data', 'orders_checked.csv'))

In [ ]:
df